keps/sig-storage/20200120-generic-data-populators.md

```diff
@@ -0,0 +1,307 @@
+
+### User Stories
+
+There are a long list of possible use cases around generic data populators, and I won't
+try to detail them all here. I will detail a few that illustrate the challenges faced by
+users and developers, but it's important to see these as a few examples among many.
+
+#### VM Images
+
+One use case was relevant to KubeVirt, where the Kubernetes `Pod` object is actually a
+VM running in a hypervisor, and the `PVC` is actually a virtual disk attached to a VM.
+It's common for virtualization systems to allows VMs to boot from disks, and for disks
+to be pre-populated with various OS images. OS images tend to be stored in external
+repositories dedicated to that purpose, often with various mechanisms for retrieving
+them efficiently that are external to Kubernetes.
+
+One way to achieve this is to represent disk images as custom resources that point to
+the image repository, and to allow creation of PVCs from these custom resources such
+that the volumes come pre-populated with the correct data. Efficient population of the
+data could be left up to a purpose-built controller that knows how to get the bits
+where they need to be with minimal I/O.
+
+#### Backup/Restore
+
+Without getting into the details of how backup/restore should be implemented, it's
+clear that whatever design one chooses, a necessary step is to have the user
+(or higher level controller) create a PVC that points to the backup they want to
+restore, and have the data appear in that volume somehow.
+
+One can imagine backups simply being a special case of snapshots, in which case the
+existing design is sufficient, but if you want anything more sophisticated, there
+will inevitably be a new object type that represents a backup. While it's arguable
+that backup should be something CSI plugins should be exclusively responsible for,
+one can also argue that generic backup tools should also exist which can backup
+and restore all kind of volumes. Those tools will be apart from CSI plugins and
+yet need a way to populate volumes with restored volumes.
+
+It's also likely that multiple backup/restore implementations will be developed,
+and it's not a good idea to pick a winner at the Kubernetes API layer. It makes
+more sense to enable developers to try different approaches by making the API allow
+restoring from various kinds of things. 
+
+### Implementation Details/Notes/Constraints
+
+As noted above, the proposal is extremely simple -- just remove the validation on
+the `DataSource` field. This raises the question of WHAT will happen when users
+put new things in that field, and HOW populators will actually work with so small
+a change.
+
+It's first important to note that only consumers of the `DataSource` field today
+are the various dynamic provisioners, most notably the external-provisioner CSI
+sidecar. If the external-provisioner sidecar sees a data source it doesn't
+understand, it simply ignores the request, which is both important for forward
```

Comment:
It seems like ignoring it is wrong - the user expected SOMETHING to happen and instead they get nothing?  It seems like a case of "operation failed successfully".

Reply:
Well the funny thing about Kubernetes is that you can do things out of order. You can create a pod that refers to a PVC that doesn't exist yet. Then later you can create the PVC, but refer to a non-existent storage class. Then later you can create the storage class that refers to a provisioner that hasn't been installed yet. Then you can install the provisioner last. Eventually the volume finally gets created and the pod starts.

I agree that generally speaking people expect the system to be installed and configured such that most requests will be satisfied immediately. Given the fail-until-success model used in Kubernetes the best thing approach I can imagine to solve this problem would be for there to be a registry of "valid" data sources types, and if a user specifies a data source not on the list, we could generate an event on the PVC warning them that the data source isn't valid yet. However it wouldn't be a good idea to fail at that point because after installing a controller, the data source could become valid and the PVC could get created. A warning would be desirable though.